### PR TITLE
Add subdomains to service url helpers

### PR DESCRIPTION
**Pull request by joe4dev**
## Motivation

While migrating the networking configuration, @simonrw [commented](https://github.com/localstack/localstack/pull/9633#discussion_r1393271424) that subdomains are commonly used but insufficiently supported by the current helpers.

It would be nice to use:
```
subdomains = f"{api_id}.execute-api"
url = f"{config.external_service_url(subdomains=subdomains)}/{stage}/{path}"
```
rather than misusing the hostname and duplicating default values.

**Why 3.0?**
We anticipate that this API will be used from LocalStack Extensions quite a bit and therefore compatibility with LocalStack >=3.0 is important.

## Changes

* Add optional `subdomains` parameter to `_service_url` helpers. This change is compatible with the existing code and primarily relevant internally within LocalStack as well as for LocalStack Extensions.
* Add parametized tests covering common scenarios
* Cleanup unused legacy variables from parametrized config tests

### CODE REVIEW

File: localstack/config.py [hunk 1]
```diff
@@ -1260,28 +1260,43 @@ def get_protocol() -> str:
...
 ) -> str:
-    """Returns a service URL to an external client used outside where LocalStack runs.
-    The configurations LOCALSTACK_HOST and USE_SSL can customize these returned URLs.
-    `host` can be used to overwrite the default for subdomains.
+    """Returns a service URL (e.g., SQS queue URL) to an external client (e.g., boto3) potentially running on another
+    machine than LocalStack. The configurations LOCALSTACK_HOST and USE_SSL can customize these returned URLs.
+    The optional parameters can be used to customize the defaults.
+    Examples with default configuration:
+    * external_service_url() == http:localhost.localstack.cloud:4566
+    * external_service_url(subdomains="s3") == http:s3.localhost.localstack.cloud:4566
```
Comment (simonrw) [inline localstack/config.py:1273]:
```suggestion
    * external_service_url() == http://localhost.localstack.cloud:4566
    * external_service_url(subdomains="s3") == http://s3.localhost.localstack.cloud:4566
```

File: tests/unit/test_config.py [hunk 2]
```diff
@@ -276,3 +266,53 @@ def test_port_out_of_range(self, port):
...
+        assert url == expected_service_url
+
+    def test_internal_service_url(self):
+        # defaults
+        assert internal_service_url() == "http://localhost:4566"
+        # subdomains
+        assert (
+            internal_service_url(subdomains="abc.execute-api")
+            == "http://abc.execute-api.localhost:4566"
+        )
```
Comment (simonrw) [inline tests/unit/test_config.py:318]: TIL that `localhost` [is a TLD](https://datatracker.ietf.org/doc/html/rfc2606#section-2)!

File: localstack/config.py [hunk 3]
```diff
@@ -1260,28 +1260,43 @@ def get_protocol() -> str:
...
+    subdomains: Optional[str] = None,
 ) -> str:
-    """Returns a service URL for internal use within where LocalStack runs. Cannot be customized
-    through LOCALSTACK_HOST because we assume LocalStack runs on the same host (i.e., localhost).
+    """Returns a service URL for internal use within LocalStack (i.e., same host).
+    The configuration USE_SSL can customize these returned URLs but LOCALSTACK_HOST has no effect.
+    The optional parameters can be used to customize the defaults.
+    Examples with default configuration:
+    * internal_service_url() == http:localhost:4566
+    * internal_service_url(port=8080) == http:localhost:8080
```
Comment (simonrw) [inline localstack/config.py:1293]:
```suggestion
    * internal_service_url() == http://localhost:4566
    * internal_service_url(port=8080) == http://localhost:8080
```

File: localstack/config.py [hunk 4]
```diff
@@ -1260,28 +1260,43 @@ def get_protocol() -> str:
...
     """
     protocol = protocol or get_protocol()
+    subdomains = f"{subdomains}." if subdomains else ""
     host = host or LOCALHOST
     port = port or GATEWAY_LISTEN[0].port
-    return f"{protocol}://{host}:{port}"
+    return f"{protocol}://{subdomains}{host}:{port}"
```
Comment (simonrw) [inline localstack/config.py:1299]: minor: do we need to guard against trying to apply subdomains if the caller specifies `host="127.0.0.1"` for example? We may want to use the IP address in some circumstances, e.g. if ipv6 causes us problems again

File: tests/unit/test_config.py [hunk 5]
```diff
@@ -67,15 +67,7 @@ def ip() -> str:
...
-    """
-    Post-v2 we are deriving
-
-    * EDGE_PORT
-
-    from GATEWAY_LISTEN. We are also ensuring the configuration behaves
-    well with LOCALSTACK_HOST, i.e. if LOCALSTACK_HOST is supplied and
-    GATEWAY_LISTEN is not, then we should propagate LOCALSTACK_HOST configuration
-    into GATEWAY_LISTEN.
+    """We are deriving GATEWAY_LISTEN and LOCALSTACK_HOST from provided environment variables.
```
Comment (simonrw) [inline tests/unit/test_config.py:70]: Thanks for updating this test :thumbsup:

**Review by simonrw — APPROVED** [review]
Looks like a great and worthwhile addition, thanks!

File: localstack/config.py (hunk 1 above)
Comment (joe4dev) [reply localstack/config.py:1273]: ups, fixed in a pending commit. Thanks.

File: localstack/config.py (hunk 3 above)
Comment (joe4dev) [reply localstack/config.py:1293]: ups, fixed in a pending commit. Thanks.

File: localstack/config.py (hunk 4 above)
Comment (joe4dev) [reply localstack/config.py:1299]: mhm, good point. Not sure whether we should be as smart as we think or wait for a use case and keep it simple for now...
We would need a reliable is_ip check (ideally for IPv4, IPv5). What do you think?

**Comment by joe4dev** [comment]
ok, merging now because the last commit only updated docstrings (i.e., no code changes)

**Pull request merged**